File: tests/verify_validate/verifier/chains/cosmos.py
import hashlib
import bech32
import ecdsa

def get_cosmos_address(private_key_hex, hrp="cosmos"):
    """
    Derives a Cosmos address from a private key hex string.
    """
    try:
        # 1. Decode Private Key
        priv_key_bytes = bytes.fromhex(private_key_hex)
        
        # 2. Get Public Key (Compressed)
        sk = ecdsa.SigningKey.from_string(priv_key_bytes, curve=ecdsa.SECP256k1)
        vk = sk.verifying_key
        pub_key_bytes = vk.to_string("compressed")
        
        # 3. SHA256(pubkey)
        sha256_hash = hashlib.sha256(pub_key_bytes).digest()
        
        # 4. RIPEMD160(sha256_hash)
        ripemd160 = hashlib.new('ripemd160')
        ripemd160.update(sha256_hash)
        raw_address = ripemd160.digest()
        
        # 5. Bech32 Encode (convert 8-bit to 5-bit first)
        five_bit_data = bech32.convertbits(raw_address, 8, 5)
        address = bech32.bech32_encode(hrp, five_bit_data)
        
        return address
        return address
    except Exception as e:
        print(f"Error deriving Cosmos address: {e}")
        return None

def verify_cosmos_key(input_key, hrp="cosmos"):
    print("\n[VERIFYING COSMOS KEY]")
    
    # Strip 0x if present
    if input_key.startswith("0x"):
        input_key = input_key[2:]
        
    if len(input_key) != 64:
        print("❌ INVALID LENGTH: 64 hex chars required (32 bytes).")
        return

    try:
        address = get_cosmos_address(input_key, hrp)
        if address:
            print("✅ VALID Private Key Verified")
            print(f"   Private Key: {input_key}")
            print(f"   Address:     {address}")
            print("\n   Note: Address depends on HRP (default 'cosmos').")
            print("")
        else:
            print("❌ Failed to derive address.")
    except Exception as e:
        print(f"❌ Key Error: {e}")

File: crates/wallet/src/cosmos.rs
use bech32::{Bech32, Hrp};
use k256::ecdsa::SigningKey;
use k256::elliptic_curve::rand_core::OsRng;
use ripemd::Ripemd160;
use sha2::{Digest, Sha256};
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use vanity_core::{Address, PrivateKey, VanityGenerator};

pub struct CosmosVanityGenerator {
    hrp: String,
    prefix: String,
    suffix: String,
    case_sensitive: bool,
}

impl CosmosVanityGenerator {
    pub fn new(hrp: &str, prefix: &str, suffix: &str, case_sensitive: bool) -> Self {
        Self {
            hrp: hrp.to_string(),
            prefix: prefix.to_string(),
            suffix: suffix.to_string(),
            case_sensitive,
        }
    }

    // ... (imports remain same)

    pub fn search(&self, found_flag: Option<Arc<AtomicU64>>) -> (PrivateKey, Address) {
        let mut csprng = OsRng;
        let p_prefix = &self.prefix;
        let p_suffix = &self.suffix;
        let p_case = self.case_sensitive;
        // Code Review: Fallback to "cosmos" is active, but we should potentially warn/error in new() if invalid.
        // For search loop, we must have a valid HRP.
        let hrp = Hrp::parse(&self.hrp).unwrap_or(Hrp::parse("cosmos").expect("valid default"));

        loop {
            if let Some(ref attempts) = found_flag {
                attempts.fetch_add(1, Ordering::Relaxed);
            }

            // 1. Generate Keypair (secp256k1 compressed)
            let signing_key = SigningKey::random(&mut csprng);
            let verifying_key = signing_key.verifying_key();
            let pubkey_bytes = verifying_key.to_sec1_bytes(); // 33 bytes compressed

            // 2. SHA256(pubkey)
            let sha256_hash = Sha256::digest(&pubkey_bytes);

            // 3. RIPEMD160(sha256_hash)
            let mut ripemd_hasher = Ripemd160::new();
            ripemd_hasher.update(sha256_hash);
            let address_bytes = ripemd_hasher.finalize();

            // 4. Encode Bech32
            // Code Review: Handle potential encoding errors gracefully or expect success for valid inputs.
            let address_str =
                bech32::encode::<Bech32>(hrp, &address_bytes).expect("bech32 encoding failed");

            // 5. Check Match
            let mut is_match = true;

            if !p_prefix.is_empty() {
                // Simplified Logic from Code Review
                let target_start = format!("{}1{}", self.hrp, p_prefix);

                let matches = if p_case {
                    address_str.starts_with(&target_start)
                } else {
                    address_str.starts_with(&target_start.to_lowercase())
                };

                if !matches {
                    is_match = false;
                }
            }

            if is_match && !p_suffix.is_empty() {
                let matches = if p_case {
                    address_str.ends_with(p_suffix)
                } else {
                    address_str.ends_with(&p_suffix.to_lowercase())
                };

                if !matches {
                    is_match = false;
                }
            }

            if is_match {
                let secret_bytes = signing_key.to_bytes();
                return (
                    PrivateKey::Cosmos(secret_bytes.into()),
                    Address::Cosmos(address_str),
                );
            }
        }
    }
}

impl VanityGenerator for CosmosVanityGenerator {
    fn generate(&self) -> (PrivateKey, Address) {
        let mut csprng = OsRng;
        let signing_key = SigningKey::random(&mut csprng);
        let verifying_key = signing_key.verifying_key();
        let pubkey_bytes = verifying_key.to_sec1_bytes();

        let sha256_hash = Sha256::digest(&pubkey_bytes);
        let mut ripemd_hasher = Ripemd160::new();
        ripemd_hasher.update(sha256_hash);
        let address_bytes = ripemd_hasher.finalize();

        let hrp = Hrp::parse(&self.hrp).unwrap_or(Hrp::parse("cosmos").unwrap());
        let address_str = bech32::encode::<Bech32>(hrp, &address_bytes).unwrap();

        let secret_bytes = signing_key.to_bytes();
        (
            PrivateKey::Cosmos(secret_bytes.into()),
            Address::Cosmos(address_str),
        )
    }
}
